### ISSUE TITLE
Should all libraries be in the same repo?

### ISSUE THREAD

**Issue by KodrAus** [elastic-rs/elastic#193]
# TODO

- [x] Replace other repos with deprecated notice (figure out how to do this so we keep tags of previous releases, but don't show history by default?)
- [x] Fix up links in docs

I used to have everything in the 1 repository here but split it out for easily automating doc uploads.

Now we've got `docs.rs` so this isn't a problem anymore.

I'm a bit worried having everything split will put people off contributing because this layer is really just gloss over `elastic_reqwest`, `elastic_types` and `elastic_responses`. That's not obvious and it's not obvious how to get code from those repos into this one.

I'm open to the idea of pulling everything together, but it would mean more churn just to undo a change that I made 6 months ago. Sunk cost etc etc.

On the other hand having this strong split has made it possible to design each piece to stand independently and kept them fairly cohesive. There are a few leaks in the way we do custom derive, but overall I think the edges are pretty clean. If everything came back together I'd like to keep that.

So the question is: Should we move everything into 1 repo? Or write up a `Contributing.md` that describes the repos you should checkout and how code moves upstream?

cc: @stephanbuys

**Comment by KodrAus** [comment]
I'm starting to lean towards pulling everything in together, but am not considering it a priority.

If it's all under the one repository, with the one build process we can get much better coverage for cross-cutting tests, integration tests, benchmarks etc that are checked on each build.

**Comment by theduke** [comment]
@KodrAus I'd second moving everything into one repo with a Cargo workspace.

**Comment by stephanbuys** [comment]
I also like consolidation, and one crate with "features".

**Comment by KodrAus** [comment]
Thanks for your insight @theduke and @stephanbuys :+1:

I'm convinced that consolidating the libraries here is a good way to go. Once we've got a basic async API (should be in the next breaking release) then I think it'll be a good idea to pull everything in and work on some decent integration tests and benchmarks.

Having everything under 1 repo suggests they'll have the same release cadence, so it'll probably be worth syncing the version numbers too. Maybe a logical next step from that is moving everything into a single crate, but I'm not totally convinced on that yet. At least it doesn't need to be done at the same time.

**Comment by KodrAus** [comment]
I've started [a project](https://github.com/orgs/elastic-rs/projects/7) to track this. I'll also work on some kind of cargo tool that uses a cargo workspace to keep version numbers in sync, or set them to local so it's easy to do releases.

**Comment by KodrAus** [comment]
We shouldn't actually need any kind of cargo tool, turns out the `path` field on dependencies in `Cargo.toml` can co-exist with `crates.io` dependencies.

Now that the async API is merged in I think it's a good time to go and pull these in before the next release. Then we can coordinate the versions going forward.

**Comment by KodrAus** [comment]
💥 I've merged all the repos in here, relicensed as dual MIT/APACHE and bumped the dependencies up to the next common version (`0.20.0`). I hadn't realised there was so much code in here 😄 There's some additional housekeeping before pushing a release to crates.io (we're waiting on `reqwest` anyways).

I'll update the OP with some next steps to track across the other repos to point back here, move issues etc. With cargo workspaces I think this change is worth the churn.
